### PR TITLE
Add puppeteer

### LINKED ISSUE #31: Add Test Framework for E2E Tests
A continuation of #17 but for E2E tests.

Possibly useful links:
https://extendomattic.wordpress.com/2019/02/20/woocommerce-extensions-e2e-tests-part-1-selecting-the-framework/

We've decided to go with Puppeteer instead of Cypress. See discussion in https://betterexperiments.wordpress.com/2020/04/23/end-to-end-testing-framework/.

### PR DESCRIPTION
**Pull request by dhurlburtusa**
Add puppeteer for our E2E testing.

Resolves #31.

## How has this been tested?

Ran `npm run build && npm run type-check && npm run format:check && npm run lint && npm run test:unit -- --coverage && npm run test:e2e` locally and everything passed.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -20,7 +35,10 @@ jobs:
                 command: npm run lint
             - run:
                 name: Unit Testing
-                command: npm run test:unit -- --coverage
+                command: npm run test:unit -- --ci --coverage
+            - run:
+                name: End-to-End Smoke Tests
```
Comment (yanirs) [inline .circleci/config.yml:40]: What's the rationale behind running only the smoke tests on CI?

File: .circleci/config.yml [hunk 2]
```diff
@@ -7,11 +7,26 @@ jobs:
...
           steps:
-            - run: npm install
+            - run:
+                name: Install Package Dependencies
+                command: npm install
+            # Copied from https://circleci.com/orbs/registry/orb/threetreeslight/puppeteer.
+            # Not using this Orb directly to prevent re-downloading Chromium which was already
+            # done in the previous step.
+            - run:
+                name: Install Headless Chrome Dependencies
```
Comment (yanirs) [inline .circleci/config.yml:22]: Not for this PR, but I think we will need to add caching of apt packages to speed up the build.

**Comment by yanirs** [comment]
Can you please add a note to the readme?

My main question is about running only smoke tests on CircleCI. It doesn't matter right now given that we don't have non-smoke tests, but ideally we should run the full test suite (unless it's impossible for some reason).

File: .circleci/config.yml (hunk 1 above)
Comment (dhurlburtusa) [reply .circleci/config.yml:40]: @yanirs I wasn't sure if we wanted to run the full e2e test suite. I'll change to run all e2e tests.

File: .circleci/config.yml (hunk 2 above)
Comment (dhurlburtusa) [reply .circleci/config.yml:22]: @yanirs Would you please clarify what you mean by "Not for this PR"?
I'll investigate how the apt packages can be cached.

Comment (yanirs) [reply .circleci/config.yml:22]: Sorry, I meant that it's not a blocker for this PR or too important at this point, just something to think about / address in the future.

Comment (dhurlburtusa) [reply .circleci/config.yml:22]: Added #48.

**Comment by dhurlburtusa** [comment]
@yanirs Okay, the CI build runs the whole E2E test suite. Do we still need something added to the readme? If not, I think this PR is ready.

**Comment by yanirs** [comment]
> @yanirs Okay, the CI build runs the whole E2E test suite. Do we still need something added to the readme? If not, I think this PR is ready.

Thanks! Yeah, please add a note under the `Testing` section before merging, along the lines of `test:unit` (which should probably be updated to exclude the reference to E2E testing).

**Pull request merged**